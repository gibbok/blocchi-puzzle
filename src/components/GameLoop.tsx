import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Action } from '@reduxjs/toolkit';
import { DetectorKeyRepeat } from './detectorKeyRepeat';

const TICK_MS = 1000;

const calcTimeClockByLevel = (base: number, level: number) => base - level * (TICK_MS / 100);

let animId = -1;

type Props = Readonly<{ level: number; detectionKeyRepeat: DetectorKeyRepeat; cb: () => void }>;

export const GameLoop = ({ level, detectionKeyRepeat, cb }: Props): JSX.Element => {
  const dispatch = useDispatch();
  const [lastTime, setLastTime] = React.useState(0);

  const loop = (time: number): void | Action => {
    const isKeyHeld = detectionKeyRepeat.get();
    const threshold = calcTimeClockByLevel(TICK_MS, level);
    const progress = time - lastTime;
    const canGameAdvance = progress >= threshold;
    if (canGameAdvance) {
      setLastTime(time);
      if (!isKeyHeld) {
        dispatch(cb());
      }
      return undefined;
    }
    animId = window.requestAnimationFrame(loop);
    return undefined;
  };

  const cleanAnimation = (id: number) => window.cancelAnimationFrame(id);

  useEffect(() => {
    animId = window.requestAnimationFrame(loop);
    return () => cleanAnimation(animId);
  });
  return <></>;
};
