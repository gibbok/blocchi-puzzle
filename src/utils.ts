import { Tetro, Board } from './types';

export const logger = (x: Tetro | Board) => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
