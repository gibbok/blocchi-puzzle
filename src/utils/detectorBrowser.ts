import { detect } from 'detect-browser';

const getPathRelease = (version: string) => version.split('.')[0];

export const detectorBrowser = (): boolean => {
  const browser = detect();
  if (browser && browser.name && browser.version) {
    const name = browser.name;
    const version = browser.version;
    const path = Number(getPathRelease(version));

    switch (name) {
      case 'chrome':
      case 'firefox':
        return path > 80;
      case 'safari':
        return path > 13;
      case 'opera':
        return path > 62;
      default:
        return false;
    }
  } else {
    return false;
  }
};
