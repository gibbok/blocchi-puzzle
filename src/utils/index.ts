export * from './data.support.test';
export * from './logger';
