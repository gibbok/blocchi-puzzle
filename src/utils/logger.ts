import { Tetro, Board } from '../game/types';

export const logger = (x: Tetro | Board): void => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
