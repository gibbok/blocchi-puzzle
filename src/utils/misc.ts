import { IO, io } from 'fp-ts/lib/IO';
import { randomInt } from 'fp-ts/lib/Random';

export const getRandomValueFromStringEnum = <T>(strEnum: T): IO<T[keyof T]> => {
  const rndInt = randomInt(0, Object.keys(strEnum).length - 1)();
  const rndValue = Object.keys(strEnum)[rndInt] as keyof T;
  return io.of(strEnum[rndValue]);
};
