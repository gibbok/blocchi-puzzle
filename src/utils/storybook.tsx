import React from 'react';
import GlobalStyle from '../components/GlobalStyle';
import styled from 'styled-components';
import { wood } from '../assets/images';

/**
 * Decorate a story with glocal styles.
 * @param story Story
 */
export const globalStylesDecorator = (story: () => unknown): JSX.Element => (
  <React.Fragment>
    <GlobalStyle />
    {story()}
  </React.Fragment>
);

const Bkg = styled.div`
  background-image: url(${wood});
  width: 100vw;
  height: 100vh;
`;

/**
 * Decorate a story with a page background.
 * @param story Story
 */
export const pageBackground = (story: () => React.ReactNode): JSX.Element => <Bkg>{story()}</Bkg>;
