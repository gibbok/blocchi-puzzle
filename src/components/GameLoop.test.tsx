import React from 'react';
import renderer from 'react-test-renderer';
import { GameLoop, loop } from './GameLoop';
import { mockStore } from '../utils';
import { Provider } from 'react-redux';

describe('<GameLoop />', () => {
  const dkr = {
    get: jest.fn(),
    set: jest.fn(),
  };
  const cb = jest.fn();
  const store = mockStore();

  it('should not render any dom', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <GameLoop level={1} detectionKeyRepeat={dkr} cb={cb} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toMatchSnapshot();
  });

  describe('loop', () => {
    const dkr = {
      get: jest.fn().mockImplementation(() => false),
      set: jest.fn(),
    };
    const setLastTimeCbMock = jest.fn();
    const cbMock = jest.fn();
    const dispatchCbMock = jest.fn().mockImplementation(() => undefined);
    it('should call dispatch and update last time, when user does not hold key', () => {
      loop(5000, 1, 1000, dkr, setLastTimeCbMock, cbMock, dispatchCbMock);
      expect(setLastTimeCbMock).toHaveBeenCalled();
      expect(cbMock).not.toHaveBeenCalled();
      expect(dispatchCbMock).toHaveBeenCalled();
    });
  });
});
