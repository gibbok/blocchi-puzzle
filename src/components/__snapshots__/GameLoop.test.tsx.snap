// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameLoop /> should not render any dom 1`] = `null`;

exports[`<GameLoop /> should not render any dom 2`] = `null`;

exports[`<GameLoop /> xxx 1`] = `null`;
